From c5125ffe5439958688e53a93459c88c33147b737 Mon Sep 17 00:00:00 2001
From: Peng Fan <peng.fan@nxp.com>
Date: Thu, 11 Oct 2018 17:33:29 +0800
Subject: [PATCH 4824/5242] MLK-19891-2 ARM64: dts: imx8qm: support hdmi in
 domu

commit  9fce57248200595228d13e8cc216dab72199dfbb from
https://source.codeaurora.org/external/imx/linux-imx.git

Support HDMI and LVDS0_CH0 output in DomU.
HDMI RX not supported, although related dts changes are in this patch.
HDMI Audio output also not supported.

Signed-off-by: Peng Fan <peng.fan@nxp.com>
Reviewed-by: Bai Ping <ping.bai@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm64/boot/dts/freescale/Makefile             |    1 +
 .../dts/freescale/fsl-imx8qm-mek-dom0-dpu2.dts     |   36 ++++++-
 .../freescale/fsl-imx8qm-mek-domu-dpu1-hdmi.dts    |   43 +++++++++
 .../dts/freescale/fsl-imx8qm-mek-domu-dpu1.dts     |    7 +-
 .../dts/freescale/fsl-imx8qm-mek-domu-hdmi.dts     |   98 ++++++++++++++++++++
 arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi  |   21 +++++
 6 files changed, 203 insertions(+), 3 deletions(-)
 create mode 100644 arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-dpu1-hdmi.dts
 create mode 100644 arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-hdmi.dts

diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index 7755291..f06c7d0 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -34,6 +34,7 @@ dtb-$(CONFIG_ARCH_FSL_IMX8QM) += fsl-imx8qm-lpddr4-arm2.dtb \
 				 fsl-imx8qm-mek-dom0-dpu2.dtb \
 				 fsl-imx8qm-mek-domu.dtb \
 				 fsl-imx8qm-mek-domu-dpu1.dtb \
+				 fsl-imx8qm-mek-domu-dpu1-hdmi.dtb \
 				 fsl-imx8qm-mek-root.dtb \
 				 fsl-imx8qm-mek-inmate.dtb \
 				 fsl-imx8qm-mek-m4.dtb \
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0-dpu2.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0-dpu2.dts
index 75bd8f7..4810f4b 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0-dpu2.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-dom0-dpu2.dts
@@ -101,6 +101,7 @@
 				SC_R_HDMI_PLL_1
 				SC_R_HDMI_I2C_0
 				SC_R_HDMI_I2S
+				SC_R_HDMI_RX
 				SC_R_SDHC_0
 				SC_R_USB_0
 				SC_R_USB_0_PHY
@@ -363,7 +364,7 @@
 		      <&usdhc1 0x12>, <&usbotg1 0x11>,
 		      <&edma01 0x10>, <&cm41 0x09>, <&pciea 0x08>,
 		      <&vpu_decoder 0x7>, <&crypto 0x6>, <&isi_0 0x5>,
-		      <&usbotg3 0x4>;
+		      <&usbotg3 0x4>, <&hdmi 0x3>;
 };
 
 &lvds_region1 {
@@ -743,3 +744,36 @@
 	#stream-id-cells = <1>;
 	iommus = <&smmu>;
 };
+
+/* hdmi */
+&hdmi {
+	xen,passthrough;
+	#stream-id-cells = <1>;
+	iommus = <&smmu>;
+	fsl,sc_rsrc_id = <SC_R_HDMI>,
+			 <SC_R_HDMI_RX>;
+};
+
+&hdmi_rx {
+	xen,passthrough;
+};
+
+&irqsteer_hdmi {
+	xen,passthrough;
+};
+
+&irqsteer_hdmi_rx {
+	xen,passthrough;
+};
+
+&i2c0_hdmi {
+	xen,passthrough;
+};
+
+&sai_hdmi_rx {
+	xen,passthrough;
+};
+
+&sai_hdmi_tx {
+	xen,passthrough;
+};
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-dpu1-hdmi.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-dpu1-hdmi.dts
new file mode 100644
index 0000000..7464149
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-dpu1-hdmi.dts
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2018 NXP
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "fsl-imx8qm-mek-domu-dpu1.dts"
+
+/ {
+	/* Sound not enabled */
+};
+
+&ldb1_phy {
+	status = "okay";
+};
+
+&ldb1 {
+	status = "okay";
+};
+
+&i2c1_lvds0 {
+	status = "okay";
+};
+
+&hdmi {
+	compatible = "fsl,imx8qm-hdmi";
+	assigned-clocks = <&clk IMX8QM_HDMI_PXL_SEL>,
+					<&clk IMX8QM_HDMI_PXL_LINK_SEL>,
+					<&clk IMX8QM_HDMI_PXL_MUX_SEL>;
+	assigned-clock-parents = <&clk IMX8QM_HDMI_AV_PLL_CLK>,
+							<&clk IMX8QM_HDMI_AV_PLL_CLK>,
+							<&clk IMX8QM_HDMI_AV_PLL_CLK>;
+	fsl,cec;
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-dpu1.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-dpu1.dts
index cda7bcd..438c04e 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-dpu1.dts
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-dpu1.dts
@@ -457,9 +457,12 @@
 /delete-node/ &i2c0_cm40;
 /delete-node/ &i2c0_cm41;
 &irqsteer_hdmi {
-	status = "disabled";
+	status = "okay";
+};
+/*/delete-node/ &i2c0_hdmi;*/
+&irqsteer_hdmi_rx {
+	status = "okay";
 };
-/delete-node/ &i2c0_hdmi;
 /delete-node/ &irqsteer_lvds1;
 /delete-node/ &flexcan1;
 /delete-node/ &flexcan2;
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-hdmi.dts b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-hdmi.dts
new file mode 100644
index 0000000..af28b24
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-mek-domu-hdmi.dts
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2018 NXP
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "fsl-imx8qm-mek-domu.dts"
+
+/*
+ * Currently we disable the audio part, since audio in domu not ready.
+ * This file is reused from fsl-imx8qm-mek-hdmi.dts.
+ */
+
+/ {
+	passthrough {
+		/*
+		sound-hdmi-tx {
+			compatible = "fsl,imx-audio-cdnhdmi";
+			model = "imx-audio-hdmi-tx";
+			audio-cpu = <&sai_hdmi_tx>;
+			constraint-rate = <48000>;
+			protocol = <1>;
+			hdmi-out;
+		};
+
+		sound-amix-sai {
+			status = "disabled";
+		};
+
+		sound-hdmi-arc {
+			compatible = "fsl,imx-audio-spdif";
+			model = "imx-hdmi-arc";
+			spdif-controller = <&spdif1>;
+			spdif-in;
+			spdif-out;
+		};
+		*/
+	};
+};
+
+&hdmi {
+	compatible = "fsl,imx8qm-hdmi";
+	assigned-clocks = <&clk IMX8QM_HDMI_PXL_SEL>,
+			  <&clk IMX8QM_HDMI_PXL_LINK_SEL>,
+			  <&clk IMX8QM_HDMI_PXL_MUX_SEL>;
+	assigned-clock-parents = <&clk IMX8QM_HDMI_AV_PLL_CLK>,
+				 <&clk IMX8QM_HDMI_AV_PLL_CLK>,
+				 <&clk IMX8QM_HDMI_AV_PLL_CLK>;
+	fsl,cec;
+	status = "okay";
+};
+
+/*
+&amix {
+	status = "disabled";
+};
+
+&sai6 {
+	status = "disabled";
+};
+
+&sai7 {
+	status = "disabled";
+};
+
+&sai_hdmi_tx {
+	assigned-clocks =<&clk IMX8QM_ACM_HDMI_TX_SAI0_MCLK_SEL>,
+			<&clk IMX8QM_AUD_PLL1_DIV>,
+			<&clk IMX8QM_AUD_ACM_AUD_PLL_CLK1_DIV>,
+			<&clk IMX8QM_AUD_ACM_AUD_REC_CLK1_DIV>,
+			<&clk IMX8QM_AUD_SAI_HDMITX0_MCLK>;
+	assigned-clock-parents = <&clk IMX8QM_AUD_ACM_AUD_PLL_CLK1_CLK>;
+	assigned-clock-rates = <0>, <768000000>, <768000000>, <768000000>, <768000000>;
+	fsl,sai-asynchronous;
+	status = "okay";
+};
+
+&sai_hdmi_rx {
+	fsl,sai-asynchronous;
+	status = "okay";
+};
+
+&spdif1 {
+	assigned-clocks =<&clk IMX8QM_AUD_PLL0_DIV>,
+			<&clk IMX8QM_AUD_ACM_AUD_PLL_CLK0_DIV>,
+			<&clk IMX8QM_AUD_ACM_AUD_REC_CLK0_DIV>;
+	assigned-clock-rates = <786432000>, <49152000>, <12288000>;
+	status = "okay";
+};
+*/
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi b/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi
index e4659d2..484889e 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8qm-xen.dtsi
@@ -192,6 +192,27 @@
 	vpu_decoder_csr: vpu_decoder_csr@0x2d080000 {
 		reg = <0x0 0x2d080000 0x0 0x1000>;
 	};
+
+	/* hdmi */
+	di_hdmi_lpcg: di_hdmi_lpcg@0x56263000 {
+		reg = <0x0 0x56263000 0x0 0x1000>;
+	};
+
+	rx_hdmi_lpcg: rx_hdmi_lpcg@0x58263000 {
+		reg = <0x0 0x58263000 0x0 0x1000>;
+	};
+
+	img_pxl_link_hdmi_lpcg: img_pxl_link_hdmi_lpcg@0x585a0000 {
+		reg = <0x0 0x585a0000 0x0 0x1000>;
+	};
+
+	aud_hdmi_rx_sai_0_lpcg: aud_hdmi_rx_sai_0_lpcg@0x59480000 {
+		reg = <0x0 0x59480000 0x0 0x1000>;
+	};
+
+	aud_hdmi_tx_sai_0_lpcg: aud_hdmi_tx_sai_0_lpcg@0x59490000 {
+		reg = <0x0 0x59490000 0x0 0x1000>;
+	};
 };
 
 /delete-node/ &edma0;
-- 
1.7.9.5

